File: source_code - v1/FUNC_Compute_values.m
% NumUE_in_SAT: 1xM
% IdClusUE: 1xU (the id cluster of each UE)
% Rho: MxU (numSAT X numUE)
% Pow: MxU (numSAT X numUE)
function [sumrate, sumpow, Rate] = FUNC_Compute_values(NumUE_in_SAT, IdClusUE, Rho, sigma_u, Pow)
numSAT = size(NumUE_in_SAT,2); numUE = sum(NumUE_in_SAT);
Rate = zeros(1, numUE);
sumpow = sum(sum(Pow));
for m = 1:numSAT
    % users in SAT m
    Um = find(IdClusUE==m);
    for iu = 1:length(Um)
        u = Um(iu);
        Rate(u) = log2(1 + Pow(m,u)*Rho(m,u)/sigma_u);
    end
end
sumrate = sum(Rate);
end

File: source_code - v1/FUNC_Compute_values.py
import numpy as np


def FUNC_Compute_values(NumUE_in_SAT, IdClusUE, Rho, sigma_u, Pow):
    numSAT = len(NumUE_in_SAT)
    numUE = int(np.sum(NumUE_in_SAT))
    rate = np.zeros(numUE, dtype=float)
    sumpow = float(np.sum(Pow))
    for m in range(1, numSAT + 1):
        um = np.where(IdClusUE == m)[0]
        for u in um:
            rate[u] = np.log2(1 + Pow[m - 1, u] * Rho[m - 1, u] / sigma_u)
    sumrate = float(np.sum(rate))
    return sumrate, sumpow, rate
